### PR TITLE
Allow iterable class objects to be unpacked

### PR DESCRIPTION
**Pull request by AlexWaygood**
Currently, mypy issues a false positive on this code, where `Foo` is iterable by virtue of having `Meta` as its metaclass:

```python
from typing import Iterator
class Meta(type):
    def __iter__(cls) -> Iterator[int]:
        yield from [1, 2, 3]

class Foo(metaclass=Meta): ...

a, b, c = Foo  # error: "Type[Foo]" object is not iterable  [misc]
reveal_type(a)  # error: Cannot determine type of "a"  [has-type]  # note: Revealed type is "Any"
```

This PR fixes that false positive.

I started working on this PR hoping to address #14782, but wasn't able to -- this PR improves the situation somewhat, but doesn't close the issue. Prior to this PR, mypy had this behaviour:

```python
from enum import Enum

class E(Enum):
    A = 1
    B = 2

A, B = E  #  error: "Type[E]" object is not iterable  [misc]
```

With this PR, mypy has this behaviour instead:

```python
from enum import Enum

class E(Enum):
    A = 1
    B = 2

A, B = E  # error: Need type annotation for "A"  [var-annotated]  # error: Need type annotation for "B"  [var-annotated]
```

This is because there's a `TypeVar` being used in [typeshed's stub for `EnumMeta.__iter__`](https://github.com/python/typeshed/blob/904407f8e4967c73bfe18fecc089cf8d22e285ce/stdlib/enum.pyi#L107): I wasn't able to figure out how to get mypy to solve the type variable in this context.

### CODE REVIEW

File: test-data/unit/check-inference.test [hunk 1]
```diff
@@ -270,6 +270,22 @@ def f() -> None:
...
+[case testInferringLvarTypesUnpackedFromIterableClassObject]
+from typing import Iterator, Type
+class Meta(type):
+    def __iter__(cls) -> Iterator[int]:
+        yield from [1, 2, 3]
+
+class Foo(metaclass=Meta): ...
+
+a, b, c = Foo
+reveal_type(a)  # N: Revealed type is "builtins.int"
```
Comment (sobolevn) [inline test-data/unit/check-inference.test:282]: Let's check `b` and `c` as well.

File: mypy/checker.py [hunk 2]
```diff
@@ -3399,6 +3399,14 @@ def check_multi_assignment(
             if len(relevant_items) == 1:
                 rvalue_type = get_proper_type(relevant_items[0])
 
+        if isinstance(rvalue_type, FunctionLike) and rvalue_type.is_type_obj():
+            ret_type = get_proper_type(rvalue_type.items[0].ret_type)
```
Comment (sobolevn) [inline mypy/checker.py:3403]: This code does not support `@overload` on `__iter__`.

For example:

```python
from typing import Iterator, Type, TypeVar, Generic, overload

T = TypeVar('T')

class Meta(type, Generic[T]):
    @overload  # special case
    def __iter__(cls: Meta[int]) -> Iterator[None]: ...
    @overload  # special case
    def __iter__(cls: Meta[T]) -> Iterator[T]: ...
```

Do we care to support this very rare case?

Comment (AlexWaygood) [reply mypy/checker.py:3403]: It looks like there are also issues for overloaded `__iter__` methods outside of the metaclass case I'm trying to tackle in this PR:

https://mypy-play.net/?mypy=latest&python=3.11&gist=fea665e2d519eeb17fc9af8ed8ee6086

Okay if I open a new issue for this and say it's out of scope here? :)

Comment (sobolevn) [reply mypy/checker.py:3403]: I am ok with it :)

Comment (AlexWaygood) [reply mypy/checker.py:3403]: I've opened this issue for the more general problem with overloaded `__iter__` methods:

- #14811

An overloaded `__iter__` method in a metaclass wouldn't be supported anyway; mypy hates generic metaclasses:

```python
from typing import AnyStr, Generic, overload, Iterator

class Meta(type, Generic[AnyStr]):
    x: AnyStr
    @overload
    def __iter__(self: Meta[str]) -> Iterator[bytes]: ...
    @overload
    def __iter__(self: Meta[bytes]) -> Iterator[str]: ...
    def __iter__(self) -> Iterator[str | bytes]:
        if isinstance(self.x, str):
            yield from [b"a", b"b", b"c"]
        yield from ["a", "b", "c"]

class Foo(metaclass=Meta[str]): ...  # error: Dynamic metaclass not supported for "Foo"  [misc]

a, b, c = Foo  # error: "Type[Foo]" object is not iterable  [misc]
reveal_type(a)  # error: Cannot determine type of "a"  [has-type]  # note: Revealed type is "Any"
```

Comment (AlexWaygood) [reply mypy/checker.py:3406]: > Okay if you leave a comment in code!

Comment added in 73f92eb

File: mypy/checker.py [hunk 3]
```diff
@@ -3385,6 +3385,9 @@ def check_multi_assignment(
         undefined_rvalue: bool = False,
     ) -> None:
         """Check the assignment of one rvalue to a number of lvalues."""
+        # TODO: Overloaded `__iter__` methods
+        # lead to unpacked variables being silently inferred as `Any`.
+        # See #14811
```
Comment (AlexWaygood) [inline mypy/checker.py:3390]:
```suggestion
```

**Comment by AlexWaygood** [comment]
I found a much better solution, which also fixes the `enum` issue

**Comment by AlexWaygood** [comment]
> ```python
> from enum import Enum
> 
> class E(Enum):
>     A = 1
>     B = 2
> 
> A, B = E  #  error: "Type[E]" object is not iterable  [misc]
> ```

- #14827

**Pull request closed without merging**